### PR TITLE
🏗 Add test runner to gulp performance

### PR DESCRIPTION
**Pull request by estherkim**
Updates to `gulp performance`:
- Adds flags for threshold and url.
- Adds a test framework. Runs test for every metric on every URL, using the threshold flag.

### CODE REVIEW

File: build-system/tasks/performance/index.js [hunk 1]
```diff
@@ -32,13 +33,17 @@ async function performance() {
   await compileScripts(urls);
   await rewriteScriptTags(urls);
   await getMetrics(urls, {headless, runs});
+  runTests();
   printReport(urls);
 }
 
 performance.description = '  Runs web performance test on current branch';
```
Comment (rsimha) [inline build-system/tasks/performance/index.js]: Nit: To make `gulp help` more readable, remove the two spaces before `performance.description` and add two spaces before each string in `performance.flags`. (It's our low-tech way of indenting flag descriptions.)

File: build-system/tasks/performance/index.js [hunk 2]
```diff
@@ -32,13 +33,17 @@ async function performance() {
...
 
 performance.flags = {
   'nobuild': 'Does not compile javascripts before running tests',
+  'threshold':
+    'Percentage above stable metrics that is passable. Number between 0 and 1',
```
Comment (rsimha) [inline build-system/tasks/performance/index.js]: For clarity, can you specify if the number is a fraction, or if it's a number in percentage points?

Also, curious: why do we limit it to the 0-1 range?

File: build-system/tasks/performance/index.js [hunk 3]
```diff
@@ -32,13 +33,17 @@ async function performance() {
...
+  'url': 'Page to test. Overrides urls set in config.json',
```
Comment (rsimha) [inline build-system/tasks/performance/index.js]: Does this mean the override is always a single URL? (Can it take a comma separated list?)

File: build-system/tasks/performance/print-report.js [hunk 4]
```diff
@@ -100,4 +100,40 @@ function printReport(urls) {
   });
 }
 
-module.exports = printReport;
+class PageMetrics {
```
Comment (rsimha) [inline build-system/tasks/performance/print-report.js]: Could you add a comment above this class to briefly explain its purpose?

File: build-system/tasks/performance/run-tests.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+const Mocha = require('mocha');
+const path = require('path');
+const TEST_SUITE_PATH = 'build-system/tasks/performance/test-suite.js';
+
+function runTests() {
+  const mocha = new Mocha();
+  mocha.addFile(path.join('./', TEST_SUITE_PATH);
+  mocha.run(failures => {
+    return new Error(failures);
```
Comment (rsimha) [inline build-system/tasks/performance/run-tests.js]: Curious: would throwing an error here mean that the task will end after the first failure is encountered? If so, does it make sense to collect failures and print them all at the end? (If not, ignore this!)

File: build-system/tasks/performance/test-suite.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const argv = require('minimist')(process.argv.slice(2));
+const loadConfig = require('./load-config');
+const {expect} = require('chai');
+const {getReport} = require('./print-report');
+// Set threshold to 1.05 by default
+const THRESHOLD = argv.threshold ? argv.threshold + 1 : 1.05;
```
Comment (rsimha) [inline build-system/tasks/performance/test-suite.js]: Maybe call this `DEFAULT_THRESHOLD`?

File: build-system/tasks/performance/index.js (hunk 3 above)
Comment (estherkim) [reply build-system/tasks/performance/index.js]: Just a URL, having more than 1 makes me think it should be using the config file.

File: build-system/tasks/performance/run-tests.js (hunk 5 above)
Comment (estherkim) [reply build-system/tasks/performance/run-tests.js]: Mocha collects all errors.

File: build-system/tasks/performance/test-suite.js (hunk 6 above)
Comment (estherkim) [reply build-system/tasks/performance/test-suite.js]: It's not the default though?

File: build-system/tasks/performance/index.js (hunk 2 above)
Comment (estherkim) [reply build-system/tasks/performance/index.js]: I thought limiting it to 200% was reasonable. I added the limit to clarify that the number is a ..fraction.. ? As in, `threshold=0.5` means the branch can be 150% above stable. Any suggestions to clarify this?

File: build-system/tasks/performance/print-report.js (hunk 4 above)
Comment (estherkim) [reply build-system/tasks/performance/print-report.js]: Done.

File: build-system/tasks/performance/test-suite.js (hunk 6 above)
Comment (rsimha) [reply build-system/tasks/performance/test-suite.js]: Yes, what you said 😃 (The intent of my comment was to move the default to its own const)

File: build-system/tasks/performance/index.js (hunk 2 above)
Comment (rsimha) [reply build-system/tasks/performance/index.js]: Gotcha. How about something like this, to make it clear that it's a floating point number?

```
Fraction by which metrics are allowed to increase. Number between 0.0 and 1.0.
```

File: build-system/tasks/performance/index.js [hunk 7]
```diff
@@ -32,13 +33,17 @@ async function performance() {
...
+  runTests();
   printReport(urls);
 }
 
-performance.description = '  Runs web performance test on current branch';
+performance.description = 'Runs web performance test on current branch';
 
 performance.flags = {
-  'nobuild': 'Does not compile javascripts before running tests',
+  'nobuild': '  Does not compile javascripts before running tests',
```
Comment (rsimha) [inline build-system/tasks/performance/index.js]: nit: While you're here, can you re-word "Does not compile javascripts..." to "Does not compile minified runtime..."?

Comment (estherkim) [reply build-system/tasks/performance/index.js]: Done

File: build-system/tasks/performance/test-suite.js (hunk 6 above)
Comment (estherkim) [reply build-system/tasks/performance/test-suite.js]: Ah ok! Done

File: build-system/tasks/performance/index.js (hunk 2 above)
Comment (estherkim) [reply build-system/tasks/performance/index.js]: Thanks! Done

**Pull request merged**